ct/reconciler: Build and commit objects using the L1 metastore

wdberkeley opened the pull request:
This changes the metastore to build and commit objects using the L1 metastore. This is necessarily
a large change, almost a rewrite. Individual commit messages have more detail.

The final commit is the big one... unfortunately it's not easy to split into meaningful, self-contained
changes.

## Backports Required

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* none

wdberkeley committed:
ct/reconciler: Add metastore dependency

This prepares the reconciler to use the metastore to manager
L1 metadata.

wdberkeley committed:
ct/reconciler: Look up tidp on attach

It makes more sense to associate a topic id partition with a cloud
topics ntp when it's attached to the reconciler, rather than each
run of the reconciler.

wdberkeley committed:
ct/reconciler: Build objects by metastore object assignment

This is a large refactor of the reconciler to use the metastore to
drive object construction and register objects in L1.
1. The reconciler processes attached leader partitions in groups based
   on common object assignment by the L1 metastore. If the object grows
   too large, some partitions are left to be reconciled in a later run.
2. The reconciler commits the resulting objects to the L1 metastore and
   updates its LRO for the partition based on its work, or based on the
   metastore's corrected offset if the metastore provides ones.
This completes the integration of L1 with the reconciler.

Unfortunately, it's basically a rewrite of the core reconciliation
logic, and it's difficult to split apart because the metastore is
involved everywhere, and good overall organization developed organically
as I rewrote the reconciliation logic.

The structure of the code makes it amenable to unit testing but I've
left that out for now. The (ignored) ducktape test still passes, which
indicates that the reconciler still uploads something to the right
prefix. This is a basic indication that object construction and the
metastore operations succeed, but further testing is needed in
follow-ups.

wdberkeley commented:
> Large diffs are not rendered by default.

😬

WillemKauf commented on src/v/cloud_topics/reconciler/reconciler.h:
```diff
@@ -153,6 +155,7 @@ class reconciler {
     data_plane_api* _data_plane;
     l1::io* _l1_io;
     cluster::metadata_cache* _metadata_cache;
+    std::unique_ptr<l1::metastore> _metastore;
```
I think `app` should own the `metastore`. I'll need it for compaction as well.

wdberkeley replied:
Hmm... Andrew suggested the reconciler own the metastore dependency. It works because the replicated metastore is a wrapper around the frontend. However, for testing purposes it'd be nice to be able to pass any implementation of metastore to the reconciler and share that metastore with other components, and the simple metastore is not a wrapper-- it owns its state. So, yeah, makes sense to instantiate a replicated metastore owned by app and pass it in here.

wdberkeley replied:
@andrwng if he gets a chance.

andrwng replied:
SGTM.

I'd originally been thinking of `metastore` as more of a "metastore client" and that there'd be a metastore client factory owned by `app`. But given the impl seems to have shaken out that main metastore is stateless, having one owned by app makes sense too

WillemKauf commented on src/v/cloud_topics/reconciler/reconciler.cc:
```diff
@@ -180,174 +232,353 @@ ss::future<> reconciler::reconciliation_loop() {
+ss::future<std::expected<reconciler::object_metadata, reconcile_error>>
+reconciler::build_object(
+  builder_context& ctx, const std::vector<attached_partition>& partitions) {
+    chunked_vector<partition_commit_info> metas;
```
essentially converting from `std::vector<>` to `chunked_vector<>` here- should all call sites be using `chunked_vector` at this point?

May also want to `.reserve()` and then `.shrink_to_fit()` after loop.

WillemKauf commented on src/v/cloud_topics/reconciler/reconciler.cc:
```diff
@@ -180,174 +232,353 @@ ss::future<> reconciler::reconciliation_loop() {
+    auto metadata_builder = _metastore->object_builder();
+    absl::flat_hash_map<l1::object_id, std::vector<attached_partition>>
+      oid_to_partitions;
+    absl::flat_hash_map<l1::object_id, builder_context> oid_to_builder;
```
unused?

wdberkeley replied:
Whoops.

WillemKauf reviewed (commented):
is there any testing of the reconciler/metastore interaction that can be done at this point?

WillemKauf commented on src/v/cloud_topics/reconciler/reconciler.h:
```diff
@@ -101,20 +114,48 @@ class reconciler {
+     * with the metastore.
      */
-    struct built_object {
+    struct object_metadata {
```
confusing name since `metastore::object_metadata` also exists.

wdberkeley replied:
Yeah, I mean it is the object metadata. I'll give it a more verbose name just to distinguish it.

dotnwat reviewed (commented):
> Unfortunately, it's basically a rewrite of the core reconciliation
logic, and it's difficult to split apart because the metastore is
involved everywhere, and good overall organization developed organically
as I rewrote the reconciliation logic.

If most of the reconciler is being rewritten, then it seems reasonable to have a single commit, because it is essentially greenfield code. But can you arrange for the diff to optimize for that? right now a large pct of the new/old code are interleaved in the rendered diff. that might be as simple as putting all new code at the end of the file. changing file name, etc... several techniques.

But it does seem like the changes could be split. For example, just taking one case: commit_object is replaced by the plural commit_objects. but the later would presumably also handle a single object, so it could be replaced in isolation?

anyway, i'm good with whichever approach you want to take.

rockwotj commented:
> is there any testing of the reconciler/metastore interaction that can be done at this point?

I really think we need to move towards this! If the only think preventing this is `cluster::partition` and the `cloud_topics::frontend` pieces we should hide them behind an interface so we can write unit tests.

also because my delete records + retention proposal is adding more stuff to the reconciler, so having testing ready would be great. If you want help with this I'm happy to collaborate.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--1c20aca3-51d3-45f4-b779-5064cdc1b184-->
<!--TEST RESULTS-->
<!--c67e8f21-01b5-4ab4-9362-649a10a54811-->
<!--TEST RESULTS-->
<!--a32ffdf0-47eb-4275-b9fd-165a7eb3548d-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72299">build#72299</a></summary>


| test_class              | test_method                         | test_arguments                                                                                                                           | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:------------------------|:------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| AuditLogTestOauth       | test_kafka_oauth                    | {"audit_transport_mode": "rpc", "authz_match": "acl"}                                                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/72299#01994f9b-fcb1-4fae-a191-92f6c0daab9f | FLAKY         | 20/21    | upstream reliability is '96.05263157894737'. current run reliability is '95.23809523809523'. drift is 0.81454 and the allowed drift is set to 50. The test should PASS  |
| SchemaEvolutionE2ETests | test_old_schema_writer              | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "produce_mode": "avro", "query_engine": "trino", "test_case": "promote_column"} | integration | https://buildkite.com/redpanda/redpanda/builds/72299#01994f9b-fcbb-4437-93fa-793f66571395 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               |
| PartitionBalancerTest   | test_recovery_mode_rebalance_finish | null                                                                                                                                     | integration | https://buildkite.com/redpanda/redpanda/builds/72299#01994f9b-fcb4-4748-9094-f2570facef43 | FLAKY         | 18/21    | upstream reliability is '85.38205980066445'. current run reliability is '85.71428571428571'. drift is -0.33223 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72504">build#72504</a></summary>


| test_class                | test_method            | test_arguments                                                                                                                       | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:--------------------------|:-----------------------|:-------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTest   | test_write             | null                                                                                                                                 | integration | https://buildkite.com/redpanda/redpanda/builds/72504#01995a01-12c1-4f08-a46a-5ac22e4ca979 | FLAKY         | 11/21    | upstream reliability is '100.0'. current run reliability is '52.38095238095239'. drift is 47.61905 and the allowed drift is set to 50. The test should PASS            |
| EndToEndCloudTopicsTxTest | test_write             | null                                                                                                                                 | integration | https://buildkite.com/redpanda/redpanda/builds/72504#01995a01-12c2-4d89-ba17-5b9554324bc9 | FLAKY         | 14/21    | upstream reliability is '100.0'. current run reliability is '66.66666666666666'. drift is 33.33333 and the allowed drift is set to 50. The test should PASS            |
| SchemaEvolutionE2ETests   | test_old_schema_writer | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "produce_mode": "avro", "query_engine": "trino", "test_case": "add_column"} | integration | https://buildkite.com/redpanda/redpanda/builds/72504#01995a00-bcde-4ae9-8f50-d4bf20c7bd4d | FLAKY         | 20/21    | upstream reliability is '98.82352941176471'. current run reliability is '95.23809523809523'. drift is 3.58543 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72549">build#72549</a></summary>


| test_class                        | test_method                 | test_arguments                  | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                      | test_history                                                                                                                                                                |
|:----------------------------------|:----------------------------|:--------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| AuditLogTestKafkaApi              | test_no_auth_enabled        | {"audit_transport_mode": "rpc"} | integration | https://buildkite.com/redpanda/redpanda/builds/72549#01995df8-5232-4774-b6fd-263af617abbb | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=AuditLogTestKafkaApi&test_method=test_no_auth_enabled                     |
| EndToEndCloudTopicsTest           | test_write                  | null                            | integration | https://buildkite.com/redpanda/redpanda/builds/72549#01995df8-5235-43b1-b75a-b38746e6db08 | FLAKY         | 11/21    | upstream reliability is '100.0'. current run reliability is '52.38095238095239'. drift is 47.61905 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=EndToEndCloudTopicsTest&test_method=test_write                            |
| VerifyConsumerOffsetsThruUpgrades | test_consumer_group_offsets | {"versions_to_upgrade": 3}      | integration | https://buildkite.com/redpanda/redpanda/builds/72549#01995df7-041a-43f1-9886-eeec303c5d11 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=VerifyConsumerOffsetsThruUpgrades&test_method=test_consumer_group_offsets |
</details>
<!--/body-->

dotnwat commented:
> is there any testing of the reconciler/metastore interaction that can be done at this point?
>> I really think we need to move towards this!

Anything L0 can provide here to help?

Lazin commented on src/v/cloud_topics/reconciler/reconciler.cc:
```diff
@@ -180,174 +232,353 @@ ss::future<> reconciler::reconciliation_loop() {
+                      next_lro, kafka::next_offset(obj_partition.last_offset));
+                }
+            }
+            partition.partition->lro = next_lro;
```
This state is not persisted anywhere, we should do something like this:

```cpp
    auto stm
      = range.partition->partition->raft()->stm_manager()->get<ctp_stm>();
    vassert(stm, "ctp_stm must be present on cloud topic partitions");

    retry_chain_node rtc(_as);
    ctp_stm_api api(rtc, stm);

    auto ret = co_await api.advance_reconciled_offset(
          next_lro);
    if (!ret.has_value()) {
          vlog(
            lg.info,
            "Failed to advance LRO to {}: {}",
            range.partition->lro,
            ret.error());
          co_return;
    }
```

dotnwat replied:
good catch. @wdberkeley did you have different/later plans for that?

wdberkeley replied:
I was planning to do it in a follow-up but it's not much code so I can throw it in as another commit on the end?

wdberkeley replied:
Yeah on second thought I'd like to handle it as a follow-up because there's both getting the LRO to start reconciliation and updating the LRO for the next reconciliation. Feels logically separate and worth thinking about on its own.

dotnwat replied:
https://redpandadata.atlassian.net/browse/CORE-13424

rockwotj commented on src/v/cloud_topics/reconciler/reconciler.cc:
```diff
@@ -180,174 +232,353 @@ ss::future<> reconciler::reconciliation_loop() {
-    auto output_stream = co_await staging_file->output_stream();
+    // Begin by creating the set of objects to be built.
+    auto metadata_builder = _metastore->object_builder();
+    absl::flat_hash_map<l1::object_id, std::vector<attached_partition>>
```
should we just start off the bat with `chunked_hash_map` and `chunked_vector`? I don't think we care that much about the performance differences here between the two.

rockwotj reviewed (commented):
Just one comment on using our chunked data structures, otherwise looks solid.

wdberkeley committed:
ct/reconciler: Container usage improvements

1. Remove unused map.
2. Reserve and shrink to fit in one instance.
3. Use chunked vectors and maps.

wdberkeley committed:
ct/reconciler: Rename object_metadata -> built_object_metadata

wdberkeley commented:
Force push to squash the metastore dependency changes into the relevant original commit. Other updates are new commits.

dotnwat commented:
@wdberkeley looks good do you want to rebase to fix the conflict for the final review pass?

wdberkeley commented:
Force push to rebase + fix conflicts with #27605

dotnwat commented on src/v/cloud_topics/reconciler/reconciler.h:
```diff
@@ -153,6 +155,7 @@ class reconciler {
     data_plane_api* _data_plane;
     l1::io* _l1_io;
     cluster::metadata_cache* _metadata_cache;
+    l1::metastore* _metastore;
```
i don't think this first commit will compile since this would be an unused private member?

wdberkeley replied:
Yeah, lemme fix that so it doesn't complicate a bisect or something.

Lazin commented on src/v/cloud_topics/reconciler/reconciler.cc:
```diff
@@ -181,191 +233,366 @@ ss::future<> reconciler::reconciliation_loop() {
+    }
+
+    if (!leaders.empty()) {
+        std::shuffle(
```
should we try to group partitions based on their properties instead of shuffling them randomly? (e.g. group compacted partitions together, or partitions that belong to the same topic)

wdberkeley replied:
Good idea but I think that's something to think about later.

dotnwat replied:
> should we try to group partitions based on their properties instead of shuffling them randomly? (e.g. group compacted partitions together, or partitions that belong to the same topic)

@wdberkeley maybe you can add a tab to the RFC to collect different optimization ideas across the cloud topics project? evgeny and i were just talking about another one related to reconciler this morning.

Lazin commented on src/v/cloud_topics/reconciler/reconciler.cc:
```diff
@@ -181,191 +233,366 @@ ss::future<> reconciler::reconciliation_loop() {
+            auto ex = object_fut.get_exception();
+            vlogl(
+              lg,
+              ssx::is_shutdown_exception(ex) ? ss::log_level::debug
```
In case of shutdown exception it makes sense to stop reconciling because the next call to `reconcile_partition` will also likely return the same error.

Lazin commented on src/v/cloud_topics/reconciler/reconciler.cc:
```diff
@@ -181,191 +233,366 @@ ss::future<> reconciler::reconciliation_loop() {
-        auto ex = upload_fut.get_exception();
+ss::future<std::expected<reconciler::object_metadata, reconcile_error>>
+reconciler::reconcile_partitions(
+  const l1::object_id& oid, const std::vector<attached_partition>& partitions) {
```
should be chunked_vector (here and in few other places)

Lazin commented on src/v/cloud_topics/reconciler/reconciler.cc:
```diff
@@ -181,191 +233,366 @@ ss::future<> reconciler::reconciliation_loop() {
+
+    // Always cleanup.
+    co_await ctx.close_builder();
+    co_await ctx.cleanup_staging();
```
is it possible for the `ctx.close_builder()` to throw the exception?

dotnwat replied:
> is it possible for the ctx.close_builder() to throw the exception?

@wdberkeley was there an answer to this?

EDIT: seems to have been addresed in a later commit

Lazin commented on src/v/cloud_topics/reconciler/reconciler.cc:
```diff
@@ -181,191 +233,366 @@ ss::future<> reconciler::reconciliation_loop() {
+          lg.error,
+          "Failed to put L1 object {}: {}",
+          oid,
+          static_cast<int>(put_result.error()));
```
should the temporary file be removed here too?

wdberkeley replied:
It is removed, a couple calls up.

dotnwat replied:
@wdberkeley maybe you can point that out? i looked up many lines i didn't see anything that looked like a removal.

Lazin replied:
It's difficult to see that this is the case. I'd prefer a general mechanism that guarantees that the files are not leaking. There is no mechanism to clean them up after the start as far as I can see so if redpanda is crashing a lot it is possible to waste all disk space.
One option is RAII think that I proposed. It is also possible to create a file, open in and immediately delete but keep the file descriptor so the actual file is not deleted from disk until the file descriptor is closed. We should also manage the disk space occupied by this scratch space (maybe it's done somewhere else, but I do not see it anywhere).

Lazin commented on src/v/cloud_topics/reconciler/reconciler.cc:
```diff
@@ -181,191 +233,366 @@ ss::future<> reconciler::reconciliation_loop() {
+                .tidp = partition.partition->tidp,
+                .base_offset = obj_partition.first_offset,
+                .last_offset = obj_partition.last_offset,
+                .max_timestamp = obj_partition.max_timestamp,
```
why there is no base timestamp?

wdberkeley replied:
The metastore doesn't track that.

rockwotj replied:
@Lazin what is base timestamp used for?

Lazin replied:
Time boundaries of adjacent segments or extents doesn't have to be aligned, right? They can overlap or there could be gaps. With timequery we need to find the earliest extent that matches the query. The base timestamp is better for this.

rockwotj replied:
> With timequery we need to find the earliest extent that matches the query. The base timestamp is better for this.

No this is incorrect. Do we do this for TS? The `first_timestamp` in a batch is exactly that: the timestamp of the first record in the batch. The `timestamp_delta` is allowed to be negative. So you can't assume the min timestamp in a batch is the first_timestamp.

There is not such thing as base timestamp without iterating over the batch.

Lazin commented on src/v/cloud_topics/reconciler/reconciler.h:
```diff
@@ -101,20 +114,48 @@ class reconciler {
+        std::unique_ptr<l1::staging_file> staging;
+        std::unique_ptr<l1::object_builder> builder;
+        size_t size_budget{max_object_size};
+
```
maybe it makes sense to add a destructor with the assertion that checks that the temporary file was deleted

Lazin replied:
another option is to hold a gate for the reconciler and schedule deletion in the background fiber using RAII

```cpp
~builder_context() {
  ssx::spawn_with_gate(parent->_gate, /*paraent is a pointer to reconciler*/, [this] {
     return cleanup_staging();
  });
}
```

so basically the pointer to the parent reconciler should be captured in the `builder_context` in order for this to work

wdberkeley committed:
ct/reconciler: Improve shutdown and cleanup

wdberkeley commented:
Force push to fix non-compiling commit prefix.

wdberkeley commented:
Force push for rebase

dotnwat commented on src/v/cloud_topics/reconciler/reconciler.cc:
```diff
@@ -181,191 +233,366 @@ ss::future<> reconciler::reconciliation_loop() {
-              partition->partition->ntp());
-            continue;
+        auto meta = co_await add_partition_to_object(ctx, partition);
+        if (meta.has_value()) {
```
nit: should an error be logged if meta doesn't have a value?

dotnwat commented on src/v/cloud_topics/reconciler/reconciler.cc:
```diff
@@ -181,191 +233,366 @@ ss::future<> reconciler::reconciliation_loop() {
+      partition->lro);
+
+    frontend fe(partition->partition, _data_plane);
+    auto reader = co_await make_reader(&fe, partition->lro, ctx.size_budget);
+    reconciliation_consumer consumer(ctx.builder.get(), partition->tidp);
+    auto metadata = co_await std::move(reader).consume(
+      std::move(consumer), model::no_timeout);
```
cool

dotnwat commented on src/v/cloud_topics/reconciler/reconciler.cc:
```diff
@@ -181,191 +233,366 @@ ss::future<> reconciler::reconciliation_loop() {
+      = add_objects_result.value().corrected_next_offsets;
+    for (const auto& obj_meta : objects) {
+        for (const auto& partition : obj_meta.partitions) {
+            auto next_lro = kafka::offset::min();
+            if (corrected_next_offsets.contains(partition.partition->tidp)) {
+                next_lro = corrected_next_offsets.at(partition.partition->tidp);
+            } else {
+                auto [first, last] = obj_meta.object_info.index.partitions
+                                       .equal_range(partition.partition->tidp);
+                for (auto it = first; it != last; ++it) {
+                    const auto& obj_partition = it->second;
+                    next_lro = std::max(
+                      next_lro, kafka::next_offset(obj_partition.last_offset));
+                }
+            }
```
we can talk about this later, but i'm surprised to see the reconciler deciding what the LRO is. IMO L1 should be the one determining what the LRO is, and the reconciler will just pass it along back to L0.

rockwotj replied:
https://github.com/redpanda-data/redpanda/blob/7f8c14fb2739b372a124aaf52bae359c9a7b1d1d/src/v/cloud_topics/level_one/metastore/metastore.h#L151-L159

rockwotj replied:
Yeah this else clause should probably not exist I'm guessing?

andrwng replied:
It's true that L1 could always return back the LRO, but I opted to have the add_response be empty in the common case because I was concerned that if this runs very frequently, the RPC cost would be non-negligible
